### PR TITLE
add python module for particle initialization

### PR DESCRIPTION
**Pull request by PrometheusPi**
This pull request adds a **python module** that allows to set up an arbitrary particle distribution to simulate e.g. an electron bunch. It also comes with a method to copy particle + field data after an initialization run with a free pusher to a new setup with e.g. a Boris pusher to take into account bunch evolution due to external fields. 

Still **to do**:

 - [x] add jupyter notebook as an example of a full initialization
 - [x] load data from h5 dump instead of checkpoint (see discussion with @psychocoderHPC in #1913)
 - [ ] add as modules inside `lib/python/picongpu/init/`
 - [ ] missing license header in new files
 - [ ] make sure all new files pass `pep8 fileName.py`

### CODE REVIEW

**Comment by ax3l** [comment]
pretty cool, shall we maybe move it to the scope of [`deppp`](https://github.com/ComputationalRadiationPhysics/deppp)?

**Comment by PrometheusPi** [comment]
That would be a good idea - I think `picongpu-tools` would be the right place because it is not just a density profile tool but allows defining the entire phase space. 

We should move it (with all other tools and modules) to such a repo if it exists.

File: src/tools/share/python/initialzeBunch.ipynb [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "collapsed": false
+   },
+   "outputs": [],
+   "source": [
+    "# define values\n",
+    "\n",
+    "# grof values\n",
```
Comment (PrometheusPi) [inline src/tools/share/python/initialzeBunch.ipynb]: typo: grid

File: src/tools/share/python/initialzeBunch.ipynb [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+    "delta_t = 1.6674254e-17\n",
+    "delta_x = 8.120995783129748e-8 * 4.0\n",
+    "delta_y = 5.0e-9\n",
+    "\n",
+    "# number of GPUs\n",
+    "GPUs = (512, 3840, 512)\n",
```
Comment (PrometheusPi) [inline src/tools/share/python/initialzeBunch.ipynb]: wrong: number of cells

File: src/tools/share/python/initialzeBunch.ipynb [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+    "\n",
+    "# iterations till PIConGPU has inialized the fields\n",
+    "N_init_timesteps = 200000\n",
+    "\n",
+    "# characteristic parameters of the electron bunch\n",
+    "mean_kinetic_energy = 15.0e6 # [eV]\n",
+    "norm_trans_emmitt_notSI = 0.5 # [mm mrad]\n",
+    "rel_energy_spread = 0.008\n",
+    "interaction_distance = 5.66e-3 + N_init_timesteps * delta_t * constants.speed_of_light # m\n",
```
Comment (PrometheusPi) [inline src/tools/share/python/initialzeBunch.ipynb]: better focus position

**Comment by ax3l** [comment]
while we are on it, let us rename the dir `modules4picongpu/` to `modules/` in one swipe (even though it existed already)

**Comment by ax3l** [comment]
> That would be a good idea - I think picongpu-tools would be the right place because it is not just a density profile tool but allows defining the entire phase space.

as discussed offline: it always starts in our case with a density profile, generates particles in 3D and then the 3V part is just modulated on top. We are doing a very similar thing for the scattering profiles (Malte, Melanie) so one could put it on a common ground with `deppp` which already has all the meta data in place before modulating a `T_e` or general `3V` on it.

**Comment by ax3l** [comment]
does it make sense to quash this PR in one or two (module, notebook) commits?

**Comment by PrometheusPi** [comment]
Yes, that would make sense - I will squash it into two commits after some (typo) fixes.

**Comment by ax3l** [comment]
admittedly, I am not the biggest fan of adding full-flavored notebooks to this repo since they are rather non-source/"compiled" and hard to review/maintain along the actual code. modules, md/reStrcuturedText code examples or a link to a gist with a ipynb with a usage example might be better...

**Comment by PrometheusPi** [comment]
@ax3l I agree - is there any better option for adding `*.ipynp` **and** having a version control on it (which gist does not support as far as I know)

**Comment by ax3l** [comment]
yep as I wrote: separate repo *or* a gist, both are versioned :)

**Comment by PrometheusPi** [comment]
@ax3l How would you generate a markdown text (with code examples) from this `*.ipynb` (as mentioned in your previous post)?

**Comment by ax3l** [comment]
If you installed [pandoc](http://pandoc.org/), you can export it directly from jupyter. This requires some non-python modules and might be hard to do on the cluster.

In order to do the same from your CLI locally where installing pandoc is no problem, just save your `*.ipynb` locally, install pandoc locally and use [`jupyter nbconvert`](https://github.com/jupyter/nbconvert). You can try if `.rst` or `.md` works best for you, both are fine for example check-in.

**Comment by ax3l** [comment]
I added you outstanding todo's in your PR description, so I don't have to look through my comments while you are still working on this PR

**Comment by ax3l** [comment]
I still think that should go in a `picongpu-utils` or `picongpu-tools` repo instead of here since it is basically openPMD file interaction, but we can talk about this when the other todo's are finished next week

**Comment by ax3l** [comment]
as in #1979, I think a good place to put those scripts is as a python module for an example.
The best example might be in `examples/Bunch/pymodules/`. Don't forget to add a `__init__.py` in this (new) dir and we might want to add it to `submitAction.sh` to be copied when found.

**Comment by PrometheusPi** [comment]
@ax3l What do you mean with using `.rst` files instead of `.ipynb`? Should I save the `.ipynb` as `.rst` and commit that file?

**Comment by ax3l** [comment]
no, instead of `.ipynb` files please only add `.py` modules. you are free to show *very short* code snippets in `README.rst` files, but you can do the same just in the module in a `__main__` section of in classes' doc strings.

File: src/tools/share/python/doc_bunchInit.md [hunk 4]
```diff
@@ -0,0 +1,219 @@
+```python
```
Comment (ax3l) [inline src/tools/share/python/doc_bunchInit.md]: this file (as `.rst` file) can go in the `docs/source/usage/workflows/` section.

File: src/tools/share/python/initialzeBunch.ipynb [hunk 5]
```diff
@@ -0,0 +1,354 @@
+{
```
Comment (ax3l) [inline src/tools/share/python/initialzeBunch.ipynb]: this file should not be added

File: src/tools/share/python/modules4picongpu/__init__.py [hunk 6]
```diff
@@ -18,3 +18,5 @@
 #
 
 from .hdf5Radiation import radiationHDF5  # NOQA
```
Comment (ax3l) [inline src/tools/share/python/modules4picongpu/__init__.py]: this file does not exist anymore and is not needed anymore

File: src/tools/share/python/modules4picongpu/bunchInit.py [hunk 7]
```diff
@@ -0,0 +1,430 @@
+import numpy as np
```
Comment (ax3l) [inline src/tools/share/python/modules4picongpu/bunchInit.py]: this module should go into `src/tools/lib/python/`

**Comment by ax3l** [comment]
@PrometheusPi what is the status of this PR? Can you please add it to the modules as commented?

**Comment by ax3l** [comment]
@PrometheusPi can you please update the PR as indicated? :)

**Comment by PrometheusPi** [comment]
@ax3l Hopefully I will find time this week to finish this.

**Comment by BeyondEspresso** [comment]
I tried out several versions of this code (serial as submitted here and parallel). I extended it with MPI communication and parallel HDF5-calls very similar to [PR: Added MPI collective I/O support](https://github.com/h5py/h5py/pull/648/files). All reading is realively fast on 16 Sandy cores on Taurus and finished 30GB within minutes. However, writing happens with a bandwidth of staggering 50kB/s per core according to `pidstat -dl 20` 😭 . So it appears at least, that there is still work left on this issue.

My question to everybody in this thread: Did anyone realize *parallel* writing with a h5py interface *and* a good bandwidth?

**Comment by ax3l** [comment]
@BeyondEspresso thanks for the observation, that's valuable feedback!

50kB/s looks less like a problem that needs to be solved immediately via parallel I/O but more like a problem in the read-write patterns used in the script.

Can you quickly profile the script and check where the actual hotspots sit? Use the `line_profiler` described in this blog post: https://www.huyng.com/posts/python-performance-analysis (an alternative is [cProfiler](https://julien.danjou.info/blog/2015/guide-to-python-profiling-cprofile-concrete-case-carbonara))

**Comment by ax3l** [comment]
@PrometheusPi can you please move the module in `lib/python/picongpu`?

**Comment by PrometheusPi** [comment]
@ax3l There are further updates to this pull request
I will close this pull request and resubmit soon with a corrected path.

**Pull request closed without merging**